File: apps/android/app/src/main/java/me/thanish/prayers/se/domain/PrayerTimeCity.kt
package me.thanish.prayers.se.domain

import android.content.Context
import androidx.datastore.preferences.core.intPreferencesKey
import me.thanish.prayers.se.R

/**
 * PrayerTimeCity is the city of a prayer time.
 */
enum class PrayerTimeCity(private val key: Int) {
    alingsas(R.string.prayers_city_alingsås),
    amal(R.string.prayers_city_åmål),
    angelholm(R.string.prayers_city_ängelholm),
    avesta(R.string.prayers_city_avesta),
    bengtsfors(R.string.prayers_city_bengtsfors),
    boden(R.string.prayers_city_boden),
    bollnas(R.string.prayers_city_bollnäs),
    boras(R.string.prayers_city_borås),
    borlange(R.string.prayers_city_borlänge),
    eksjo(R.string.prayers_city_eksjö),
    enkoping(R.string.prayers_city_enköping),
    eskilstuna(R.string.prayers_city_eskilstuna),
    eslov(R.string.prayers_city_eslöv),
    falkenberg(R.string.prayers_city_falkenberg),
    falkoping(R.string.prayers_city_falköping),
    filipstad(R.string.prayers_city_filipstad),
    flen(R.string.prayers_city_flen),
    gallivare(R.string.prayers_city_gällivare),
    gavle(R.string.prayers_city_gävle),
    gislaved(R.string.prayers_city_gislaved),
    gnosjo(R.string.prayers_city_gnosjö),
    goteborg(R.string.prayers_city_göteborg),
    halmstad(R.string.prayers_city_halmstad),
    haparanda(R.string.prayers_city_haparanda),
    harnosand(R.string.prayers_city_härnösand),
    hassleholm(R.string.prayers_city_hässleholm),
    helsingborg(R.string.prayers_city_helsingborg),
    hogsby(R.string.prayers_city_högsby),
    horby(R.string.prayers_city_hörby),
    hudiksvall(R.string.prayers_city_hudiksvall),
    hultsfred(R.string.prayers_city_hultsfred),
    jokkmokk(R.string.prayers_city_jokkmokk),
    jonkoping(R.string.prayers_city_jönköping),
    kalmar(R.string.prayers_city_kalmar),
    kalrshamn(R.string.prayers_city_kalrshamn),
    karlskoga(R.string.prayers_city_karlskoga),
    karlskrona(R.string.prayers_city_karlskrona),
    karlstad(R.string.prayers_city_karlstad),
    katrineholm(R.string.prayers_city_katrineholm),
    kiruna(R.string.prayers_city_kiruna),
    koping(R.string.prayers_city_köping),
    kristianstad(R.string.prayers_city_kristianstad),
    kristinehamn(R.string.prayers_city_kristinehamn),
    laholm(R.string.prayers_city_laholm),
    landskrona(R.string.prayers_city_landskrona),
    lessebo(R.string.prayers_city_lessebo),
    lidkoping(R.string.prayers_city_lidköping),
    linkoping(R.string.prayers_city_linköping),
    ludvika(R.string.prayers_city_ludvika),
    lulea(R.string.prayers_city_luleå),
    lund(R.string.prayers_city_lund),
    lysekil(R.string.prayers_city_lysekil),
    malmo(R.string.prayers_city_malmö),
    mariestad(R.string.prayers_city_mariestad),
    marsta(R.string.prayers_city_märsta),
    mellerud(R.string.prayers_city_mellerud),
    mjolby(R.string.prayers_city_mjölby),
    monsteras(R.string.prayers_city_mönsterås),
    munkedal(R.string.prayers_city_munkedal),
    nassjo(R.string.prayers_city_nässjö),
    norrkoping(R.string.prayers_city_norrköping),
    norrtalje(R.string.prayers_city_norrtälje),
    nybro(R.string.prayers_city_nybro),
    nykoping(R.string.prayers_city_nyköping),
    nynashamn(R.string.prayers_city_nynäshamn),
    orebro(R.string.prayers_city_örebro),
    ornskoldsvik(R.string.prayers_city_örnsköldsvik),
    oskarshamn(R.string.prayers_city_oskarshamn),
    ostersund(R.string.prayers_city_östersund),
    oxelosund(R.string.prayers_city_oxelösund),
    pajala(R.string.prayers_city_pajala),
    pitea(R.string.prayers_city_piteå),
    ronneby(R.string.prayers_city_ronneby),
    saffle(R.string.prayers_city_säffle),
    sala(R.string.prayers_city_sala),
    savsjo(R.string.prayers_city_sävsjö),
    simrishamn(R.string.prayers_city_simrishamn),
    skara(R.string.prayers_city_skara),
    skelleftea(R.string.prayers_city_skellefteå),
    skovde(R.string.prayers_city_skövde),
    soderhamn(R.string.prayers_city_söderhamn),
    sodertalje(R.string.prayers_city_södertälje),
    solleftea(R.string.prayers_city_sollefteå),
    solvesborg(R.string.prayers_city_sölvesborg),
    stockholm(R.string.prayers_city_stockholm),
    strangnas(R.string.prayers_city_strängnäs),
    sundsvall(R.string.prayers_city_sundsvall),
    tierp(R.string.prayers_city_tierp),
    tranemo(R.string.prayers_city_tranemo),
    trelleborg(R.string.prayers_city_trelleborg),
    trollhattan(R.string.prayers_city_trollhättan),
    uddevalla(R.string.prayers_city_uddevalla),
    ulricehamn(R.string.prayers_city_ulricehamn),
    umea(R.string.prayers_city_umeå),
    uppsala(R.string.prayers_city_uppsala),
    vanersborg(R.string.prayers_city_vänersborg),
    varberg(R.string.prayers_city_varberg),
    varnamo(R.string.prayers_city_värnamo),
    vasteras(R.string.prayers_city_västerås),
    vastervik(R.string.prayers_city_västervik),
    vaxjo(R.string.prayers_city_växjö),
    vetlanda(R.string.prayers_city_vetlanda),
    vimmerby(R.string.prayers_city_vimmerby),
    visby(R.string.prayers_city_visby),
    ystad(R.string.prayers_city_ystad);

    /**
     * getLabel returns the name of the city with i18n.
     */
    fun getLabel(context: Context): String {
        return context.getString(key)
    }

    /**
     * ✄ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
     */

    companion object {
        /**
         * STORE_KEY is the MMKV key for storing the current prayer city.
         */
        private val STORE_KEY = intPreferencesKey("PrayerTimeCity")


        /**
         * set sets the current prayer city.
         */
        fun set(context: Context, city: PrayerTimeCity) {
            setIntegerSync(context, STORE_KEY, city.ordinal)
        }

        /**
         * get returns the current prayer city.
         */
        fun get(context: Context): PrayerTimeCity {
            val index = getIntegerSync(context, STORE_KEY, stockholm.ordinal)
            return entries[index]
        }
    }
}

File: apps/iphone/Bönetider/Bönetider/Domain/PrayerTimeCity.swift
//
//  PrayerTimeCity.swift
//  Bönetider
//
//  Created by Thanish Nizam on 2025-01-31.
//

import SwiftUI

// PrayerTimeCity is the city the prayer time is calculated for.
// This value van be used on SwiftUI views using the snippet given below.
//
//   @AppStorage(PrayerTimeCity.key)
//   private var city: PrayerTimeCity = PrayerTimeCity.standard
//
// To update the stored value, set the PrayerTimeCity.current value.
//
//   PrayerTimeCity.current = PrayerTimeCity.uppsala
//
enum PrayerTimeCity: String, CaseIterable, Identifiable {
  case alingsas, amal, angelholm, avesta, bengtsfors, boden, bollnas, boras, borlange
  case eksjo, enkoping, eskilstuna, eslov, falkenberg, falkoping, filipstad, flen
  case gallivare, gavle, gislaved, gnosjo, goteborg, halmstad, haparanda, harnosand
  case hassleholm, helsingborg, hogsby, horby, hudiksvall, hultsfred, jokkmokk, jonkoping
  case kalmar, kalrshamn, karlskoga, karlskrona, karlstad, katrineholm, kiruna, koping
  case kristianstad, kristinehamn, laholm, landskrona, lessebo, lidkoping, linkoping
  case ludvika, lulea, lund, lysekil, malmo, mariestad, marsta, mellerud, mjolby, monsteras
  case munkedal, nassjo, norrkoping, norrtalje, nybro, nykoping, nynashamn, orebro
  case ornskoldsvik, oskarshamn, ostersund, oxelosund, pajala, pitea, ronneby
  case saffle, sala, savsjo, simrishamn, skara, skelleftea, skovde, soderhamn, sodertalje
  case solleftea, solvesborg, stockholm, strangnas, sundsvall, tierp, tranemo, trelleborg
  case trollhattan, uddevalla, ulricehamn, umea, uppsala, vanersborg, varberg, varnamo
  case vasteras, vastervik, vaxjo, vetlanda, vimmerby, visby, ystad
  
  // Key used to store notification offset on UserDefaults
  static let key: String = "PrayerTimeCity"
  
  // The default value to use before a value is picked by the user
  static let standard: PrayerTimeCity = .stockholm
  
  // The current prayer time city value stored in UserDefaults
  static var current : PrayerTimeCity {
    get {
      let storedValue = UserDefaults.standard.string(forKey: key)
      return Self(rawValue: storedValue ?? "") ?? PrayerTimeCity.standard
    }
    set { UserDefaults.standard.set(newValue.id, forKey: key) }
  }
  
  // MARK: - Computed Properties
  
  // Unique and machine friendly identifier used for matching
  var id: String { self.rawValue }
  
  // Localized string of the prayer time cuty
  var label: String {
    switch self {
      case .alingsas:
        return String(localized: "prayers_city_alingsas")
      case .amal:
        return String(localized: "prayers_city_amal")
      case .angelholm:
        return String(localized: "prayers_city_angelholm")
      case .avesta:
        return String(localized: "prayers_city_avesta")
      case .bengtsfors:
        return String(localized: "prayers_city_bengtsfors")
      case .boden:
        return String(localized: "prayers_city_boden")
      case .bollnas:
        return String(localized: "prayers_city_bollnas")
      case .boras:
        return String(localized: "prayers_city_boras")
      case .borlange:
        return String(localized: "prayers_city_borlange")
      case .eksjo:
        return String(localized: "prayers_city_eksjo")
      case .enkoping:
        return String(localized: "prayers_city_enkoping")
      case .eskilstuna:
        return String(localized: "prayers_city_eskilstuna")
      case .eslov:
        return String(localized: "prayers_city_eslov")
      case .falkenberg:
        return String(localized: "prayers_city_falkenberg")
      case .falkoping:
        return String(localized: "prayers_city_falkoping")
      case .filipstad:
        return String(localized: "prayers_city_filipstad")
      case .flen:
        return String(localized: "prayers_city_flen")
      case .gallivare:
        return String(localized: "prayers_city_gallivare")
      case .gavle:
        return String(localized: "prayers_city_gavle")
      case .gislaved:
        return String(localized: "prayers_city_gislaved")
      case .gnosjo:
        return String(localized: "prayers_city_gnosjo")
      case .goteborg:
        return String(localized: "prayers_city_goteborg")
      case .halmstad:
        return String(localized: "prayers_city_halmstad")
      case .haparanda:
        return String(localized: "prayers_city_haparanda")
      case .harnosand:
        return String(localized: "prayers_city_harnosand")
      case .hassleholm:
        return String(localized: "prayers_city_hassleholm")
      case .helsingborg:
        return String(localized: "prayers_city_helsingborg")
      case .hogsby:
        return String(localized: "prayers_city_hogsby")
      case .horby:
        return String(localized: "prayers_city_horby")
      case .hudiksvall:
        return String(localized: "prayers_city_hudiksvall")
      case .hultsfred:
        return String(localized: "prayers_city_hultsfred")
      case .jokkmokk:
        return String(localized: "prayers_city_jokkmokk")
      case .jonkoping:
        return String(localized: "prayers_city_jonkoping")
      case .kalmar:
        return String(localized: "prayers_city_kalmar")
      case .kalrshamn:
        return String(localized: "prayers_city_kalrshamn")
      case .karlskoga:
        return String(localized: "prayers_city_karlskoga")
      case .karlskrona:
        return String(localized: "prayers_city_karlskrona")
      case .karlstad:
        return String(localized: "prayers_city_karlstad")
      case .katrineholm:
        return String(localized: "prayers_city_katrineholm")
      case .kiruna:
        return String(localized: "prayers_city_kiruna")
      case .koping:
        return String(localized: "prayers_city_koping")
      case .kristianstad:
        return String(localized: "prayers_city_kristianstad")
      case .kristinehamn:
        return String(localized: "prayers_city_kristinehamn")
      case .laholm:
        return String(localized: "prayers_city_laholm")
      case .landskrona:
        return String(localized: "prayers_city_landskrona")
      case .lessebo:
        return String(localized: "prayers_city_lessebo")
      case .lidkoping:
        return String(localized: "prayers_city_lidkoping")
      case .linkoping:
        return String(localized: "prayers_city_linkoping")
      case .ludvika:
        return String(localized: "prayers_city_ludvika")
      case .lulea:
        return String(localized: "prayers_city_lulea")
      case .lund:
        return String(localized: "prayers_city_lund")
      case .lysekil:
        return String(localized: "prayers_city_lysekil")
      case .malmo:
        return String(localized: "prayers_city_malmo")
      case .mariestad:
        return String(localized: "prayers_city_mariestad")
      case .marsta:
        return String(localized: "prayers_city_marsta")
      case .mellerud:
        return String(localized: "prayers_city_mellerud")
      case .mjolby:
        return String(localized: "prayers_city_mjolby")
      case .monsteras:
        return String(localized: "prayers_city_monsteras")
      case .munkedal:
        return String(localized: "prayers_city_munkedal")
      case .nassjo:
        return String(localized: "prayers_city_nassjo")
      case .norrkoping:
        return String(localized: "prayers_city_norrkoping")
      case .norrtalje:
        return String(localized: "prayers_city_norrtalje")
      case .nybro:
        return String(localized: "prayers_city_nybro")
      case .nykoping:
        return String(localized: "prayers_city_nykoping")
      case .nynashamn:
        return String(localized: "prayers_city_nynashamn")
      case .orebro:
        return String(localized: "prayers_city_orebro")
      case .ornskoldsvik:
        return String(localized: "prayers_city_ornskoldsvik")
      case .oskarshamn:
        return String(localized: "prayers_city_oskarshamn")
      case .ostersund:
        return String(localized: "prayers_city_ostersund")
      case .oxelosund:
        return String(localized: "prayers_city_oxelosund")
      case .pajala:
        return String(localized: "prayers_city_pajala")
      case .pitea:
        return String(localized: "prayers_city_pitea")
      case .ronneby:
        return String(localized: "prayers_city_ronneby")
      case .saffle:
        return String(localized: "prayers_city_saffle")
      case .sala:
        return String(localized: "prayers_city_sala")
      case .savsjo:
        return String(localized: "prayers_city_savsjo")
      case .simrishamn:
        return String(localized: "prayers_city_simrishamn")
      case .skara:
        return String(localized: "prayers_city_skara")
      case .skelleftea:
        return String(localized: "prayers_city_skelleftea")
      case .skovde:
        return String(localized: "prayers_city_skovde")
      case .soderhamn:
        return String(localized: "prayers_city_soderhamn")
      case .sodertalje:
        return String(localized: "prayers_city_sodertalje")
      case .solleftea:
        return String(localized: "prayers_city_solleftea")
      case .solvesborg:
        return String(localized: "prayers_city_solvesborg")
      case .stockholm:
        return String(localized: "prayers_city_stockholm")
      case .strangnas:
        return String(localized: "prayers_city_strangnas")
      case .sundsvall:
        return String(localized: "prayers_city_sundsvall")
      case .tierp:
        return String(localized: "prayers_city_tierp")
      case .tranemo:
        return String(localized: "prayers_city_tranemo")
      case .trelleborg:
        return String(localized: "prayers_city_trelleborg")
      case .trollhattan:
        return String(localized: "prayers_city_trollhattan")
      case .uddevalla:
        return String(localized: "prayers_city_uddevalla")
      case .ulricehamn:
        return String(localized: "prayers_city_ulricehamn")
      case .umea:
        return String(localized: "prayers_city_umea")
      case .uppsala:
        return String(localized: "prayers_city_uppsala")
      case .vanersborg:
        return String(localized: "prayers_city_vanersborg")
      case .varberg:
        return String(localized: "prayers_city_varberg")
      case .varnamo:
        return String(localized: "prayers_city_varnamo")
      case .vasteras:
        return String(localized: "prayers_city_vasteras")
      case .vastervik:
        return String(localized: "prayers_city_vastervik")
      case .vaxjo:
        return String(localized: "prayers_city_vaxjo")
      case .vetlanda:
        return String(localized: "prayers_city_vetlanda")
      case .vimmerby:
        return String(localized: "prayers_city_vimmerby")
      case .visby:
        return String(localized: "prayers_city_visby")
      case .ystad:
        return String(localized: "prayers_city_ystad")
    }
  }
}
